asciidoc/courses/llm-chatbot-typescript/modules/2-chains/lessons/2-answer-generation/lesson.adoc

```diff
@@ -0,0 +1,115 @@
+
+In this challenge, you must modify the `initGenerateAnswerChain()` function in link:{repository-blob}/main/{lab-file}[`{lab-file}`^] to return a chain that instructs the LLM to generate an answer to a question based on the context provided.
+The chain should not be concerned with where that information comes from.
+
+The chain will accept the following input:
+
+.Chain Input
+[source,typescript]
+----
+include::{repository-raw}/{branch}/{lab-solution}[tag=interface]
+----
+
+The chain will need to:
+
+1. Format a prompt based on the input to the chain.  The prompt template must include `{question}` and `{context}`
+2. Pass the formatted prompt to the LLM
+3. Convert the output to a string
+
+lab::Open `{lab-filename}`[]
+
+
+[TIP]
+.Modifying Functions
+====
+Each challenge in this course will follow the same structure.
+You will be given a helper function that will accept the appropriate arguments.
+Your job will be to complete the implementation, using the arguments passed to the function.
+
+Later in this course, as you build out the agent, you will use these functions to create instances of the chains and register them as tools.
+====
+
+
+== Create a Prompt Template
+
+Inside the `initGenerateAnswerChain()` function, use the `PromptTemplate.fromTemplate()` method to create a new prompt template.
+Use the following prompt as the first parameter.
+
+.Prompt
+[source]
+----
+include::{repository-raw}/{branch}/prompts/{prompt-filename}[]
+----
+
+Due to the nature of semantic search, the documents passed by the application to this prompt may not fully answer the question.
+Therefore this prompt instructs the LLM to do its best to generate an answer based on the input, or respond with "I don't know" if it cannot answer the question.
+
+Later in the course, you will build a chain that will provide a definitive answer.
+
+Your code should resemble the following:
+
+
+.Prompt Template
+[source,typescript,indent=0]
+----
+include::{repository-raw}/{branch}/{lab-solution}[tag=prompt, indent=0]
+----
+
+== Create the Runnable Sequence
+
+Use the `RunnableSequence.from()` method to create a new chain.
+The chain should pass the prompt to the LLM passed as a parameter, then format the response as a string using a new instance of the `StringOutputParser`.
+
+[source,typescript]
+----
+include::{repository-raw}/{branch}/{lab-solution}[tag=sequence, indent=0]
+----
+
+Use the `return` keyword to return the chain from the function.
+
+== Working Solution
+
+.Click here to reveal the fully-implemented `answer-generation.chain.ts`
+[%collapsible]
+====
+[source,js,indent=0]
+----
+include::{repository-raw}/{branch}/{lab-solution}[tag=function]
+----
+====
+
+== Using the Chain
+
+You will be able to initialize and run the chain in your application with the following code:
+
+[source,typescript]
+----
+include::{repository-raw}/{branch}/{lab-solution}[tag=usage]
```

Comment:
This code is already included in the starter code - https://github.com/neo4j-graphacademy/llm-chatbot-typescript/blob/main/src/modules/agent/chains/answer-generation.chain.ts#L24

Should it be? Or should there just be a placeholder?

Reply:
It was originally in this file and not in the solution.  I must have copied it over to the chains.  I don't think it does any harm being in both places